### DISCUSSION TITLE
GDB on VS-Code with Lima

### DISCUSSION THREAD

**Discussion by GiacomoPaltrinieri** [lima-vm/lima#1433] [Q&A] [answered]
Hello everyone, my university requires the use of GDB as a debugger for C++. However, I am unable to install it on my MacBook M1, either natively or through emulation. As a workaround, I have installed Lima and installed GDB on the virtual machine. Does anyone know how to configure VS-Code to integrate Lima and GDB? Thank you.

Please let me know if you have any further requirements.

**Comment by afbjorklund** [comment, marked as answer]
I think you would use something like https://code.visualstudio.com/docs/cpp/pipe-transport, to do remote debugging...

Assuming that you have already configured the compiler, so that it would either run with `lima` or cross-compile to Linux ?

-----

It could be useful to have a whole example for Lima, like WSL:

https://code.visualstudio.com/docs/remote/remote-overview

![](https://code.visualstudio.com/assets/docs/remote/remote-overview/architecture.png)

You can set up ssh config to the Lima VM, using `limactl show-ssh`

Then VS Code will install the VS Code Server, on the lima instance.

**Reply by GiacomoPaltrinieri** [reply]
It worked!! Thanks a lot!!!!!!!!

I ended up using VS-Code with the remote SSH extension:
![image](https://user-images.githubusercontent.com/80061200/226702502-21f4a04a-2340-4664-ad51-56aec267c80e.png).

**Reply by afbjorklund** [reply]
Parts of the things on the WSL page are also true, like the file sharing and the port forwarding... Not the other details, though.

* https://code.visualstudio.com/docs/remote/ssh

* https://code.visualstudio.com/docs/remote/wsl

Wonder if it would be possible to contribute a Lima page ?

**Reply by GiacomoPaltrinieri** [reply]
I'm sorry, English is not my main language but i don't get what you're talking about here: "Wonder if it would be possible to contribute a Lima page ?"

**Reply by afbjorklund** [reply]
As in:

* https://code.visualstudio.com/docs/remote/lima :-)

**Reply by GiacomoPaltrinieri** [reply]
page not found :(

**Reply by afbjorklund** [reply]
No, it needs to be written first... There are some hints at the bottom of that page, how to contribute to their documentation.

Could be something, for later. Similar for the dev containers below, they do mention it - but only Rancher Desktop and colima.

**Reply by GiacomoPaltrinieri** [reply]
I haven't really documented anything yet, but i could try to describe the steps i used if needed.

**Reply by afbjorklund** [reply]
I am trying to compile some more extensive documentation, as part of the "Lima Manual" project...

https://github.com/afbjorklund/lima-man (an empty placeholder, from https://github.com/afbjorklund/lima-gui/issues/7)

**Reply by GiacomoPaltrinieri** [reply]
oh, it would be very helpful.
If i can help somehow let me know, I will do my best!

**Comment by afbjorklund** [comment]
If you don't want to install everything on the VM, you can use Dev Containers:

https://code.visualstudio.com/docs/devcontainers/containers

But it requires that you use `template://docker` rather than the default nerdctl*

https://code.visualstudio.com/remote/advancedcontainers/docker-options

----

\* There is an open feature request, to make it (nerdctl) available as an option:

* https://github.com/microsoft/vscode-remote-release/issues/6014

**Reply by afbjorklund** [reply]
Possibly one could make it limp along, by teaching [nerdctld](https://github.com/afbjorklund/nerdctld) some new tricks.

Like in the https://github.com/afbjorklund/nerdctld/tree/main/lima examples

**Reply by afbjorklund** [reply]
Hmm, seems like it is only using the Docker CLI anyway - not the Docker API ?

Then it should be enough with setting `nerdctl.lima` as the path to "Docker"....

**Reply by afbjorklund** [reply]
Nope, that did not go very far... Needs to be bug-compatible with Docker, anyway.

```
[105406 ms] Start: Run: nerdctl.lima version --format {{.Server.APIVersion}}
[105542 ms] time="2023-03-21T22:11:20Z" level=fatal msg="template: :1:9: executing \"\" at <.Server.APIVersion>: can't evaluate field APIVersion in type *dockercompat.ServerVersion"
```
